app/src/dex/router/route_and_fill.rs

```diff
@@ -93,42 +86,45 @@ pub trait RouteAndFill: StateWrite + Sized {
             lambda_2_2: unfilled_2,
         };
 
+        // TODO: how does this work when there are trades in both directions?
+        // Won't that mix up the traces? Should the SwapExecution be indexed by
+        // the _DirectedTradingPair_?
```

Comment:
@zbuc I noticed this and was a bit confused, don't we want to be recording two distinct executions for each pair per block? otherwise the traces would get mixed up between them, right? Or?

Reply:
I think you're right. That does seem wrong. It seems like we need to execute a batch swap per trading direction.